core/src/main/kotlin/net/corda/core/Futures.kt

```
@@ -0,0 +1,47 @@
+package net.corda.core
+
+import com.google.common.util.concurrent.AbstractFuture
+import com.google.common.util.concurrent.ListenableFuture
+import net.corda.core.utilities.loggerFor
+import org.slf4j.Logger
+
+/**
+ * Each future is passed into the given handler as soon as it becomes done.
+ * By default invocations of the handler do not run concurrently, pass in null for [lock] to allow them to do so.
+ * The handler should throw [ThenContext.thenAgain] if it does not want to affect the returned future.
+ * Otherwise the result of the handler is copied into the returned future.
+ * If the handler errors and the returned future is already done, the error is logged (unless it's thenAgain).
+ * If the handler returns normally and the returned future is already done, the handler value is silently discarded.
+ */
+fun <T> Iterable<ListenableFuture<*>>.then(lock: Any? = Any(), handler: ThenContext.(ListenableFuture<*>) -> T): ListenableFuture<T> {
```

I've fixed the re-entrant case now, so the handler only runs when the result future isn't done (with tests). The plan for this code (in another PR) is to implement a replacement for allAsList (that doesn't hide exceptions/hangs) on top of it, so the future that wins won't always be the first one, so for that reason I'm keeping then() as the name.

The most general case I'm thinking of is
```
val f1 = SettableFuture.create<Boolean>()
val f2 = SettableFuture.create<Boolean>()
val f3 = SettableFuture.create<Boolean>()
val g = listOf(f1, f2, f3).then {
    // Implement short-circuiting conjunction or disjunction here.
}
```
but not in this PR.